### ISSUE TITLE
Skipping deploy due to sync error: copying files: didn't sync any files

### ISSUE THREAD

**Issue by pre** [GoogleContainerTools/skaffold#4246]
### TL;DR

Skaffold Sync depends on the current namespace in $KUBECONFIG. When you develop locally with `skaffold dev` and switch the current namespace in $KUBECONFIG as part of the dev flow, Skaffold Sync becomes unreliable. 

This issue has been reported earlier but issues were just closed without a fix:
- https://github.com/GoogleContainerTools/skaffold/issues/3685
- https://github.com/GoogleContainerTools/skaffold/issues/1668

Details and a workaround: 
https://github.com/GoogleContainerTools/skaffold/issues/1668#issuecomment-595752550

### Expected behavior

Skaffold Sync should be able to copy files regardless of what happens to be the current namespace in $KUBECONFIG. 

### Actual behavior

Skaffold Sync will fail if artefact is deployed in a different namespace than which is defined in $KUBECONFIG as the current namespace.

### Information

- Skaffold version: 1.10.1 (and older)
- Operating system: MacOS
- Contents of skaffold.yaml:

```yaml
apiVersion: skaffold/v2alpha2
kind: Config
build:
  local:
    push: false
  artifacts:
    - image: lolcat
      context: lolcat
      sync:
        manual:
          - src: "app/**/*"
            dest: /

```

### PROBLEM: Namespace in $KUBECONFIG affects Skaffold Sync
Skaffold sync will *only* work when the Kubernetes namespace in $KUBECONFIG, at the time of starting Skaffold, matches the namespace of the artifact (application) being deployed. 

The critical moment is *when Skaffold starts*. After Skaffold is running, I can change my current namespace in $KUBECONFIG and Skaffold sync works neatly. But, if the namespace in $KUBECONFIG is different from the hot-sync-artifact's namespace *when Skaffold starts*, then sync will fail with error `WARN[0032] Skipping deploy due to sync error: copying files: didn't sync any files` .

For Skaffold Sync to work, the namespace in $KUBECONFIG *must* either be undefined or it must equal to the hot-sync-artifact's namespace. 
### EXAMPLE: Namespace undefined
Without namespace being defined, Skaffold Sync works for artifacts in any namespace:
```
apiVersion: v1
clusters:
- cluster:
    certificate-authority: /Users/pre/.minikube/ca.crt
    server: https://192.168.64.100:8443
  name: rex
contexts:
- context:
    cluster: rex
    user: rex
  name: rex
current-context: rex
kind: Config
preferences: {}
users:
- name: rex
  user:
    client-certificate: /Users/pre/.minikube/client.crt
    client-key: /Users/pre/.minikube/client.key
```

## EXAMPLE: Namespace defined
Here namespace is defined and Skaffold Sync will fail for all artifacts (applications) which are not deployed in `dev` namespace:
```
apiVersion: v1
clusters:
- cluster:
    certificate-authority: /Users/pre/.minikube/ca.crt
    server: https://192.168.64.100:8443
  name: rex
contexts:
- context:
    cluster: rex
    namespace: dev
    user: rex
  name: rex
current-context: rex
kind: Config
preferences: {}
users:
- name: rex
  user:
    client-certificate: /Users/pre/.minikube/client.crt
    client-key: /Users/pre/.minikube/client.key
```

### Workaround

1. Reset the current namespace in $KUBECONFIG
2. Run `skaffold dev` without a namespace being defined in $KUBECONFIG 
3. Create a subshell which switches back to the original namepace after Skaffold has started.

```bash
_dev() {
  local current_ns; current_ns="$(kubectl config view --minify --output 'jsonpath={..namespace}')"
  kubectl config set-context --current --namespace=
  (sleep 3 && kubectl config set-context --current --namespace="${current_ns}" ) &
    
  skaffold dev
}
```

Best things are achieved with a decent amount of `sleep`.

**Comment by briandealwis** [comment]
I haven't tried to reproduce your issue yet, but have you tried using Skaffold's `--kube-context` and `--kubeconfig` options to pin to a specific context?

      --kube-context='': Deploy to this Kubernetes context
      --kubeconfig='': Path to the kubeconfig file to use for CLI requests.

**Comment by pre** [comment]
I tested passing `--kube-context` and `--kubeconfig` to  Skaffold. They have no effect in this problem (Skaffold was already using the correct Kubeconfig with the correct kubecontext).

* $KUBECONFIG points to correct file and is set by the script which runs Skaffold.
* Both examples above represent a full Kubeconfig.
* $KUBECONFIG has only one context which is defined by Minikube (full example above).
 
Skaffold Sync not functioning is caused solely by `context.namespace` being defined in $KUBECONFIG _when Skaffold starts_.

**Comment by briandealwis** [comment]
Oh I see.  Skaffold uses `kubectl` under the hood for many operations: you can see these commands by running `skaffold ... -v debug`.  Each invocation will use the value of the kube-context as defined at that time.

Frequently changing your Kubernetes context definition seems to me to be against the purpose of a Kubernetes context, and increases operational risk.  Instead, define separate Kubernetes contexts to capture your different uses.  I have two contexts, one with a namespace and the other without:
```
$ kubectl config get-contexts
CURRENT   NAME         CLUSTER   AUTHINFO  NAMESPACE
          dev-lab      gke_xxx   gke_xxx
*         my-lab       gke_xxx   gke_xxx   bdealwis
```

Closing as there's nothing that we can do here.

**Comment by pre** [comment]
> Frequently changing your Kubernetes context definition seems to me to be against the purpose of a Kubernetes context, and increases operational risk.

It seems that this issue report was not understood correctly.

To summarise:
* This problem report is about `skaffold dev`.
* The cluster is Minikube running on developer's personal machine (not a shared cluster).
* Skaffold is using a dedicated $KUBECONFIG provided by Minikube. 
* There is only one Kubernetes context in $KUBECONFIG used by Skaffold (the one provided Minikube).
* Developer is using a dedicated $KUBECONFIG with a single context (the one provided by Minikube).

As such, there is no operational risk - it's not about operations, it's about development. 

If this problem is ignored and marked as #wontfix, then it seems the mitigation would be to document properly, that `skaffold dev` needs a $KUBECONFIG of its own. 

What this means is, you'll always need to wrap `skaffold dev` in a shell script, which either unsets the `context.namespace` (example above) or makes a copy of $KUBECONFIG generated by Minikube just for Skaffold.

**Comment by pre** [comment]
fyi @dgageot  - earlier you seemed interested about fixing this issue and I felt that you understood that there is a problem. Please see the TL;DR section and re-evaluate whether this problem is something that should be fixed or not.

The issue is that frequently developers just bang their head on the wall wondering why Skaffold Sync fails with error `Skipping deploy due to sync error: copying files: didn't sync any files`. 

The current behaviour of `skaffold dev` in unexpected and causes a lot of wasted working hours in the long run.

**Comment by dgageot** [comment]
I'll see why I can do

**Comment by tejal29** [comment]
@pre Sorry for the delay in any updates. 

> The current behaviour of skaffold dev in unexpected and causes a lot of wasted working hours in the long run.

Totally agree. I was wondering what we could do to fix this issue.
Skaffold uses `kubectl cp <pod>--namespace <podNs> -c` command to copy the sync file. 
Ideally, if even you switch context within skaffold dev iterations, `--namepsace` flag should point the pod in the deployed namespace.
I saw an old bug where if kubectl context namespace is set, `--namespace` value is not respected https://github.com/kubernetes/client-go/issues/288 and kubernetes/kubernetes#46299
However, this bug got fixed in 2017. I am wondering if this could be related.

**Comment by mr-bjerre** [comment]
I'm having this issue as well if I'm setting namespace in `kustomization.yaml`. Then the sync doesn't work AND I don't get any stdout to my console. 

Any fixes available?

**Comment by pre** [comment]
There’s no fix, but I’ve been using the workaround I posted in the issue description. It works well, although it’d be nice if skaffold sync would not get broken when the namespace is changed.

**Comment by omninonsense** [comment]
I've made a PR to (hopefully) address this. However, in case the PR gets rejected—it is a bit shabby—here is a more permanent workaround for minikube users that doesn't require hacking/wrapping/aliasing anything. 

When starting minikube (for the first time or consecutively) you can pass it `--namespace=`. This has a similar effect as the workaround mentioned in https://github.com/GoogleContainerTools/skaffold/issues/4246#issue-623203593.

The added benefit of doing it this way is that minikube remember this for future invocations of `minikube start`, since it will update the profile config (it sets `.KubernetesConfig.Namespace` in `~/.minikube/profiles/{profile}/config.json` to `""`). So next time you run it, it will still attempt to patch the kubeconfig, but it will not set the namespace (or it will remove it if it's there).

**Comment by nkubala** [comment]
would anyone here be able to verify this is still an issue? I believe this was fixed through #5487 but I don't have a good repro.

**Comment by hemakshis** [comment]
Thanks @omninonsense I started my minikube cluster with `minikube start --namespace=<my_namespace>` and it worked. I had been banging my head on this issue for 2 days.

**Comment by omninonsense** [comment]
Hi @nkubala! Unfortunately, I'm still getting this with skaffold v1.23.0 🙁 

```
WARN[2720] Skipping deploy due to sync error: copying files: didn't sync any files
```

**Comment by mantelllo** [comment]
Hello, @nkubala! The same issue with v1.25.0. Fortunately, the workaround still works :slightly_smiling_face:
```
minikube start --namespace=<my_namespace>
```

**Comment by hambone01** [comment]
Same issue with v1.28.0

**Comment by jhiester** [comment]
I'm also having this issue with v1.28.1
